### PR TITLE
bpo-35018: Sax parser provides no user access to lexical handlers.

### PR DESCRIPTION
**Pull request by ghost**
This PR contains the functionality that is needed to fix [bpo-6686](https://bugs.python.org/issue6686) and [bpo-9371](https://bugs.python.org/issue9371), namely the LexicalHandler class which makes it easy to access XML comments and the document encoding. Those PR's depend on this issue.

https://bugs.python.org/issue35018

### CODE REVIEW

**Comment by jfgossage** [comment]
It was my dyslexia - sorry. The correct number is bpo-9371.

**Comment by tirkarthi** [comment]
It's okay. It seems somehow the bot links the PR with ticket numbers in the PR comment. I don't know if it's intentional though. Thanks for the clarification.

**Review by serhiy-storchaka — CHANGES_REQUESTED** [review]
Please edit the title and the description of this PR.

**Comment by jfgossage** [comment]
I have made the requested changes; please review again.

**Review by serhiy-storchaka — DISMISSED** [review]

File: Lib/test/test_sax.py [hunk 1]
```diff
@@ -5,6 +5,7 @@
                     SAXException, SAXReaderNotAvailable, SAXParseException
 import unittest
 from unittest import mock
+from Lib.pickle import FALSE
```
Comment (serhiy-storchaka) [inline Lib/test/test_sax.py]: What relation does SAX have to pickle?

File: Lib/xml/sax/handler.py [hunk 2]
```diff
@@ -11,11 +11,12 @@
 
 version = '2.0beta'
 
-#============================================================================
+# ============================================================================
```
Comment (serhiy-storchaka) [inline Lib/xml/sax/handler.py]: Unrelated change.

File: Lib/xml/sax/handler.py [hunk 3]
```diff
@@ -225,7 +226,7 @@ class EntityResolver:
     implementing this interface, then register the object with your
     Parser, the parser will call the method in your object to
     resolve all external entities. Note that DefaultHandler implements
-    this interface with the default behaviour."""
+    this interface with the default behavior."""
```
Comment (serhiy-storchaka) [inline Lib/xml/sax/handler.py]: Unrelated change. Aren't both spellings correct?

File: Misc/NEWS.d/next/Library/2018-11-03-21-21-32.bpo-35018.z1vnri.rst [hunk 4]
```diff
@@ -0,0 +1,4 @@
+Add the LexicalHandler class that is present in other SAX XML
+implementations. The plumbing is already supported by Ptyhon so this simply
+adds the poecelain that makes it easy for users of the Python Sax parser to
```
Comment (serhiy-storchaka) [inline Misc/NEWS.d/next/Library/2018-11-03-21-21-32.bpo-35018.z1vnri.rst]: "poecelain"?

**Review by serhiy-storchaka — COMMENTED** [review]
Just few nitpicks.

File: Misc/NEWS.d/next/Library/2018-11-03-21-21-32.bpo-35018.z1vnri.rst (hunk 4 above)
Comment (tirkarthi) [reply Misc/NEWS.d/next/Library/2018-11-03-21-21-32.bpo-35018.z1vnri.rst]: Also I think it's Python not Ptyhon

**Comment by tirkarthi** [comment]
1. I am not a native speaker so I will defer to someone on the general tone of the documentation and the type of spelling.
2. Many of the code predates PEP8 and the general consensus is that unless you need to change it PEP8 related changes reduces the usefulness of git history that for example to track a comment you need to use git blame and changing the whitespace adds another step while reading git history for the line. Style guidelines are generally enforced for new code or code that you happen to change but making the old code PEP8 compliant around the change that you need to make is produces little gain and it's avoided.

**Comment by jfgossage** [comment]
I have made the requested changes; please review again .

**Comment by jfgossage** [comment]
@serhiy-storchaka Is it possible to mark issues bpo-6686 and bpo-9371 as depending on this issue bpo-35018. Both need the LexicalHandler to implement the fix and I don't have the authority to do this myself.

**Comment by serhiy-storchaka** [comment]
1. In any case this change is out of scope of this issue. Open a separate issue for fixing words in British spelling.

2. You should use PEP 8 for new code, and can use it for modified code, but left unrelated code untouched. And consistency with other code in the file can be more important than following some PEP 8 rules. PEP 8 explicitly allows this.

File: Lib/test/test_sax.py [hunk 5]
```diff
@@ -1339,6 +1340,169 @@ def test_nsattrs_wattr(self):
...
+                             len(self.specified_comment))
+            self.assertEqual(' {} '.format(self.specified_comment[i]),
+                             self.comments[i])
+
+
+class CDATAHandlerTest(unittest.TestCase):
+    """This is implemented as a separate class since CDATA sections in XML
+    cannot appear within elements defined within a DTD. Hence this test does
+    not use a DTD."""
+    def __init__(self, *args, **kwargs):
```
Comment (scoder) [inline Lib/test/test_sax.py:1447]: Doing something in `TestCase.__init__()`, especially setting up test data, is a very unusual thing to do. Please use the normal set-up methods instead.

File: Lib/xml/sax/handler.py [hunk 6]
```diff
@@ -340,3 +341,57 @@ def resolveEntity(self, publicId, systemId):
...
+
+    This handler is used to obtain lexical information about an XML
+    document, that is, information about how the document was encoded
+    (as opposed to what it contains, which is reported to the
+    ContentHandler), such as comments and CDATA marked section
+    boundaries.
+
+    To set the LexicalHandler of an XMLReader, use the setProperty
+    method with the property identifier
+    'http://xml.org/sax/handlers/LexicalHandler'."""
```
Comment (scoder) [inline Lib/xml/sax/handler.py]: This link gives a 404 for me.

File: Lib/xml/sax/handler.py [hunk 7]
```diff
@@ -340,3 +341,57 @@ def resolveEntity(self, publicId, systemId):
...
+
+    def xmlDecl(self, version, encoding, standalone):
```
Comment (scoder) [inline Lib/xml/sax/handler.py:359]: This method is not part of the `LexicalHandler` interface.
http://www.saxproject.org/apidoc/org/xml/sax/ext/LexicalHandler.html

**Review by scoder — CHANGES_REQUESTED** [review]
I agree with the request to exclude unrelated changes from this feature PR.

**Comment by csabella** [comment]
@jgossage, please address the code review requests.  Thanks!

**Comment by csabella** [comment]
I'm going to close this for now as the original PR author has not been active for awhile.  This can be re-opened once work is ready to continue.  Thanks!
